airflow/gcp/hooks/pubsub.py

```diff
@@ -77,47 +103,106 @@ def publish(self, project: str, topic: str, messages: List[Dict]) -> None:
         :type messages: list of PubSub messages; see
             http://cloud.google.com/pubsub/docs/reference/rest/v1/PubsubMessage
         """
-        body = {'messages': messages}
-        full_topic = _format_topic(project, topic)
-        request = self.get_conn().projects().topics().publish(  # pylint: disable=no-member
-            topic=full_topic, body=body)
+        publisher = self.get_conn()
+        topic_path = PublisherClient.topic_path(project, topic)  # pylint: disable=no-member
+
+        # TODO validation of messages
+        self.log.info("Publish %d messages to topic (path) %s", len(messages), topic_path)
         try:
-            request.execute(num_retries=self.num_retries)
-        except HttpError as e:
+            for message in messages:
+                publisher.publish(
+                    topic=topic_path,
+                    data=message.get("data", b''),
+                    **message.get('attributes', {})
+                )
+        except GoogleAPICallError as e:
             raise PubSubException(
-                'Error publishing to topic {}'.format(full_topic), e)
+                'Error publishing to topic {}'.format(topic_path), e)
+
+        self.log.info("Published %d messages to topic (path) %s", len(messages), topic_path)
 
-    def create_topic(self, project: str, topic: str, fail_if_exists: bool = False) -> None:
+    def create_topic(
+        self,
+        project: str,
+        topic: str,
+        fail_if_exists: bool = False,
+        labels: Optional[Dict[str, str]] = None,
```

Comment:
According to integration guide we should add label with Airflow version to indicate that a resource was created with the use of Airflow. For example:
https://github.com/PolideaInternal/airflow/blob/6a66ece8cae8c8d209efefcde2639dca8919e975/airflow/gcp/hooks/gcs.py#L449-L451

To obtain version use `from airflow.version import version` ;)

Reply:
Good idea. I will add it.